Initial Api Plumbing

Allcharles opened the pull request:
This is just an initial outlook on the solution to #78, DO NOT MERGE. Unit tests will fail spectacularly until I do a full rewrite.

Potential improvements:

  - Replace `allPath` and `idPath` with `CommonApiPaths` from previous solution. This would make it possible to define custom paths in the constructor instead of the special case method

Allcharles committed:
Converted project and site services

Updated project and site services to use new abstract api

Allcharles committed:
Added project/site function declarations

Allcharles committed:
Created StandardApi class

Allcharles committed:
Split User Service

Allcharles committed:
Updated user service references

atruskie committed:
Rejigged API services

The following classes should be correct and considered canonical:
- src\app\services\baw-api\projects.service.ts
- src\app\services\baw-api\sites.service.ts
- src\app\services\baw-api\api-common.ts
- src\app\services\baw-api\base-api.service.ts

All other classes are likely broken!

Allcharles commented on package.json:
```diff
@@ -102,7 +102,7 @@
     "tslint": "^5.20.1",
     "tslint-angular": "^3.0.2",
-    "typescript": "^3.7.4",
+    "typescript": "^3.8.0-dev.20200201",
```
Not a huge fan of using dev typescript versions. Currently 3.7.4 isn't even officially supported and is currently stopping us from upgrading the build-angular dependency. I'm also seeing a large list of dependencies complaining that they do not support this version of typescript.

atruskie replied:
Reverting to stable is fine

Allcharles commented on src/app/helpers/stringTemplate/stringTemplate.ts:
```diff
@@ -1,3 +1,5 @@
+import { type } from "os";
```
Unused import

Allcharles commented on src/app/services/baw-api/base-api.service.ts:
```diff
@@ -35,12 +35,56 @@ export abstract class BawApiService {
+      } else {
+        return [new _new(response.data)];
+      }
+    };
```
Not sure why these are implemented in the constructor

atruskie replied:
Ah, that was a tricky bug. 

Reactive requires pure functions. Just pulling a function off of the service class (i.e. `this.handleResponse`) risks rebinding `this`. We need `this` though so we can get access to that `_new` property.

By assigning arrow functions to the values in the constructor, we close over `_new` instead and do not need to rely on `this`.

Allcharles committed:
App builds

atruskie commented on src/app/services/baw-api/api-common.ts:
```diff
@@ -83,30 +91,46 @@ export interface ApiUpdate<M extends AbstractModel, P extends any[]> {
+/**
+ * Api class with only the show route enabled
+ */
+export abstract class ShowApi<M extends AbstractModel, P extends any[]>
```
So you probably don't need to create a child abstract class for every combination of API methods that may be implemented in a service. It was my intention that one would just implement the interface they needed.

The abstract class here is there really to just combine a whole wack of type signatures into a nice bundle.

Still, if you find this construct useful - go for it.

Allcharles committed:
Cleanup sites unit tests

Allcharles committed:
Working sites service unit tests

Allcharles committed:
Added site filter tests

Allcharles committed:
Added metadata to models

Allcharles committed:
Added http method unit tests

Updated BawApiService unit tests to properly test http request methods

Allcharles committed:
Added api method unit tests

Added unit tests for BawApiService dealing with list, filter, show, create, update, and destroy functions

Allcharles committed:
Removed api-common.spec files

Allcharles committed:
Added ShallowSitesService unit tests

Allcharles committed:
Added UserService unit tests

Allcharles committed:
Added project unit tests

Allcharles committed:
Added account unit tests

Allcharles committed:
Re-enabled test suite

Allcharles committed:
Added SecurityService unit tests

Allcharles committed:
Changed create/update function signatures

Allcharles committed:
Revert "Changed create/update function signatures"

This reverts commit 9a96150af378278c1fed1aff4ce5e0cf3eda25d5.

atruskie commented on src/app/models/AudioRecording.ts:
```diff
@@ -29,40 +29,32 @@ export interface AudioRecordingInterface {
+  public readonly dataLengthBytes?: number;
+  public readonly status?: "ready" | "uploading" | "corrupt";
+  public readonly createdAt?: DateTimeTimezone;
+  public readonly updatedAt?: DateTimeTimezone;
```
This is a good change. 

I think generally we'll have more property getters and setters defined though.

Hopefully we can use those to convert types (like strings to dates) during serialization

atruskie commented on src/app/models/Project.ts:
```diff
@@ -15,68 +15,56 @@ import { AbstractModel } from "./AbstractModel";
       ? DateTime.fromISO(project.createdAt as string, {
           setZone: true
         })
       : defaultDateTimeTimezone;
```
What purpose does `setZone` have?

Allcharles replied:
setZone makes the luxon date object accept accept timezone data. By default, luxon and javascript dates ignore timezones

atruskie replied:
Ah goodo

atruskie commented on src/app/services/baw-api/account.service.spec.ts:
```diff
@@ -8,214 +8,378 @@ import { Subject } from "rxjs";
+  const shouldNotFail = () => {
+    fail("Service should not produce an error");
+  };
+
```
these seem sufficiently general enough to be moved to a common test helpers class

atruskie commented on src/app/services/baw-api/account.service.spec.ts:
```diff
@@ -8,214 +8,378 @@ import { Subject } from "rxjs";
+      });
+    }
+
+    it("should handle empty response", fakeAsync(() => {
```
why `fakeAsync`? why not just use the `async` keyword?

Allcharles replied:
I prefer `fakeAsync` because it has all of the functionality of `async` with additional control. `async` actually waits for the timer to finish, whereas `fakeAsync` allows me to kill the timer making tests run faster.

atruskie commented on src/app/services/baw-api/account.service.spec.ts:
```diff
@@ -8,214 +8,378 @@ import { Subject } from "rxjs";
+      }, shouldNotFail);
+
+      tick(100);
+    }));
```
these tests seem redundant - what are they testing?

Allcharles replied:
The difference is that "should handle response with random user model id" uses the User model to get the id, whereas "should handle response with random user id" passes the id directly. Its mostly a check that the id is being read correctly from both types of inputs using an id value which is not 1 (which i use as a standard placeholder id)

atruskie replied:
Ah fair enough.

atruskie commented on src/app/services/baw-api/account.service.spec.ts:
```diff
@@ -8,214 +8,378 @@ import { Subject } from "rxjs";
+
+      tick(100);
+    }));
+  });
```
Generally, this test file is a lot simpler. Well done.

I'm not sure we need to test every combination you're testing here, especially for every service. But for now, good as is.

atruskie commented on src/app/services/baw-api/base-api.service.spec.ts:
```diff
@@ -1,68 +1,855 @@
+
+    httpMethods.forEach(httpMethod => {
+      describe(httpMethod.functionName, () => {
+        it("should create patch request", () => {
```
```suggestion
        it(`should create ${ httpMethod } request`, () => {
```

atruskie commented on src/app/services/baw-api/base-api.service.spec.ts:
```diff
@@ -1,68 +1,855 @@
+        it("should handle error response", () => {
+          const response = { meta: errorMeta, data: null } as ApiResponse<
+            MockModel
+          >;
```
you're line length linting rules seem a bit aggressive

Allcharles replied:
Took a crack at fixing it, Prettier (my formatter) has options to expand the line length but its ignoring whatever i set to ill have to fix it another time

atruskie commented on src/app/services/baw-api/base-api.service.ts:
```diff
@@ -36,56 +37,66 @@ export abstract class BawApiService<T extends AbstractModel> {
+    config: AppConfigService,
+    @Inject(STUB_CLASS_BUILDER) classBuilder: new (_: object) => T
+  ) {
+    this.apiRoot = config.getConfig().environment.apiRoot;
```
TODO: remove later

atruskie commented on src/app/services/baw-api/base-api.service.ts:
```diff
@@ -36,56 +37,66 @@ export abstract class BawApiService<T extends AbstractModel> {
+          model => new classBuilder(this.addMetadata(model, response.meta))
+        );
+      } else {
+        return [this.handleSingleResponse(response)];
```
inline this. This is so simple we don't need to abstract into handle single response

Allcharles replied:
I believe you abstracted this to prevent issues with the value of `this` being overwritten

atruskie replied:
No, I mean literally that line:
```suggestion
        return [new classBuilder(this.addMetadata(model, response.meta)];
```

atruskie commented on src/app/services/baw-api/base-api.service.ts:
```diff
@@ -36,56 +37,66 @@ export abstract class BawApiService<T extends AbstractModel> {
       if (response.data instanceof Array) {
-        throw new Error("Received an array of API results when only a single result was expected");
+        return response.data.map(
+          model => new classBuilder(this.addMetadata(model, response.meta))
```
I would tag with metadata after the new class has been constructed.

Object.assign semantics may make this method work, but it is clearer if the final type get tagged with metadata.

Have you considered adding a method on `AbstractModel` that deals with metadata?

atruskie commented on src/app/services/baw-api/base-api.service.ts:
```diff
@@ -36,56 +37,66 @@ export abstract class BawApiService<T extends AbstractModel> {
+  public retrieveMetadata(model: AbstractModel): Meta {
+    return model[this.metaKey];
+  }
+
```
move implementation to AbstractModel

atruskie commented on src/app/services/baw-api/base-api.service.ts:
```diff
@@ -197,18 +239,26 @@ export abstract class BawApiService<T extends AbstractModel> {
+   */
+  private addMetadata(model: object, meta: Meta): object {
+    return { ...model, [this.metaKey]: meta };
+  }
```
move this to AbstractModel

atruskie commented on src/app/services/baw-api/mock/projectsMockService.ts:
```diff
@@ -1,29 +1,39 @@
-export class MockProjectsService extends MockApiCommon<Project> {
-  public getProject() {
-    return new Subject();
+export class MockProjectsService extends StandardApi<Project, []> {
```
I wonder if there is a more general way to create mock services this?

It seems like a lot of work to create a mock service for every class.

atruskie commented on src/app/services/baw-api/projects.service.spec.ts:
```diff
@@ -7,24 +7,70 @@ import { RouterTestingModule } from "@angular/router/testing";
+  const shouldNotFail = () => {
+    fail("Service should not produce an error");
+  };
+
```
see the previous comment about abstracting out common test helpers

atruskie reviewed (approved):
Generally: the destroy action may or may not return the object that was just destroyed. So change all `destroy` return types from `null` to `M | null`.

A very good set of changes. A lot of the unit test code is easier to read and understand. 💯

We're at the point where you have enough direction, so go ahead and merge in when you've made changes and continue the bulk work on master.

Allcharles committed:
Fixed initial review comments

Allcharles committed:
Updated destroy return type

Changed destroy return type from `null` to `M | void`.

Allcharles committed:
Address pr review

Fixes comment: https://github.com/QutEcoacoustics/workbench-client/pull/82#discussion_r375647896

Allcharles committed:
Abstracted service mocks

Relates to comment: https://github.com/QutEcoacoustics/workbench-client/pull/82#discussion_r375649391

Allcharles committed:
Added json conversion for models

Allcharles committed:
Fixed MockModels

Allcharles committed:
Proper implementation of toJSON

Allcharles committed:
Renamed baw service files

- Added `.service` to api interceptor
- Renamed BawApiService file

Allcharles committed:
Merge branch 'master' into api-plumbing

Allcharles committed:
Fix merged unit tests

Allcharles committed:
Cleanup menu unit tests
